### ISSUE TITLE
Cannot Flash Keyboard

### ISSUE THREAD

**Issue by Lite5h4dow** [Dygmalab/Bazecor#291]
**Describe the bug**
Cannot seem to update firmware

**To Reproduce**
1. Follow flash instructions
2. bazecore stops responding
3. raise stops responding
4. flash doesn't fail or succeed

**Expected behavior**
flash to complete

**Screenshots**
![image](https://user-images.githubusercontent.com/18059545/138461540-11a37ac7-5cb1-4803-ae75-be4e3998964a.png)

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Bazecor Version: 0.3.4 (keyboard firmware: 0.3.3)

**Comment by vicnil** [comment]
Solved. Have to run it in sudo and with option no-sandbox. That is:
sudo ./Bazecor-1.0.0beta12.AppImage --no-sandbox


    OS: Manjaro Linux
    Bazecor Version: 1.0.0beta12

**Comment by nsrderooy** [comment]
@vicnil That is not a "solved", that is a workaround, like I stated before it works on older linux versions without sudo.

**Comment by vicnil** [comment]
@nsrderooy Yeah, I should have expressed myself differently, especially since it doesn't help Windows users (like OP). Edited my post. 

I was unaware that it works on older Linux versions without sudo. Due to the actions the Bazecor takes when flashing the keyboard, maybe this is something that can't simply be fixed in Bazecor since safety measures in the OS blocks it?

**Comment by AlexDygma** [comment]
Hey @nsrderooy @vicnil ! 

It seems something has changed on the latest linux kernel versions if you are having problems with using Bazecor on linux, we fixed this same issue a couple of months from now, so we will have to check exactly was has changed in the recent updates to solve this issue.

**Comment by AlexDygma** [comment]
So i have checked this in a `Linux vdf-135 5.10.0-9-amd64 SMP Debian 5.10.70-1 (2021-09-30) x86_64 GNU/Linux` and it asked me
![Cutoff](https://user-images.githubusercontent.com/70891884/144258217-e285be55-0c58-4c1d-9d48-91ce2267cb0f.png)
To allow it to install the Udev rules like in previous versions, take into account that this popup stays behind Bazecor main window due to being opened before the main window starts, so it should appear in the background, I'm looking into having it keep the focus so no one can miss it.

Did this window appear to you? did you accept the installation of the Udev rules? if you did I will need an environment to reproduce your error so I can understand why Udev rules are no longer enough to make it work.

**Comment by AlexDygma** [comment]
Hey @nsrderooy , @vicnil !

I added to the development branch some code that forces the popup to stay on top of the main window, so now it doesn't appear until the window finishes loading and it does so on top, so I hope that with this change it will be easier to use.

**Comment by vicnil** [comment]
@AlexDygma I tried moving the Bazecor window but couldn't find anything behind it (or is it supposed to come up not at opening the Bazecor but at some other stage?). I'm on i3 windows manager, so it could be that it makes the pop-up harder for me to find. It will be interesting to test the new version!

**Comment by nsrderooy** [comment]
@AlexDygma Thanks, but the window is not there for me (probably because the udev rules were already loaded).

```
SUBSYSTEMS=="usb", ATTRS{idVendor}=="1209", ATTRS{idProduct}=="2201", GROUP="users", MODE="0666"
```

Also just a remark on this rule, maybe it's an idea to ask the user if this is a multi-user system, as now you are giving the `users` group permission (actually the `GROUP` part is useless because the mode is `0666` so it's world writable).

On my system the `users` group is empty;

```
getent group users
users:x:100:
```

if you were to ask if the system is a multi-user system and the person answers no you can change the udev rule to;
```
SUBSYSTEMS=="usb", ATTRS{idVendor}=="1209", ATTRS{idProduct}=="2201", OWNER="${USERNAME}", MODE="0600"
```

The way it's currently setup apache can write to my keyboard (maybe not a good idea).

**Comment by AlexDygma** [comment]
> @AlexDygma I tried moving the Bazecor window but couldn't find anything behind it (or is it supposed to come up not at opening the Bazecor but at some other stage?). I'm on i3 windows manager, so it could be that it makes the pop-up harder for me to find. It will be interesting to test the new version!

Here you have it https://drive.google.com/file/d/1zij4I7AVAYjDa4wf1FnsNofPgQiL6B5h/view?usp=sharing

**Comment by AlexDygma** [comment]
> @AlexDygma Thanks, but the window is not there for me (probably because the udev rules were already loaded).
> 
> ```
> SUBSYSTEMS=="usb", ATTRS{idVendor}=="1209", ATTRS{idProduct}=="2201", GROUP="users", MODE="0666"
> ```
> 
> Also just a remark on this rule, maybe it's an idea to ask the user if this is a multi-user system, as now you are giving the `users` group permission (actually the `GROUP` part is useless because the mode is `0666` so it's world writable).
> 
> On my system the `users` group is empty;
> 
> ```
> getent group users
> users:x:100:
> ```
> 
> if you were to ask if the system is a multi-user system and the person answers no you can change the udev rule to;
> 
> ```
> SUBSYSTEMS=="usb", ATTRS{idVendor}=="1209", ATTRS{idProduct}=="2201", OWNER="${USERNAME}", MODE="0600"
> ```
> 
> The way it's currently setup apache can write to my keyboard (maybe not a good idea).

So @nsrderooy , did the udev rules sove your flashing issues on the keyboard? if not, do they appear every time you try to flash? if this is the case, toggle on the developer console in preferences (or press Ctrl + Shift + I) and try the flashing procedure again to see the output error you get, if it's a timing thing (not detecting bootloader mode when de keyboard enters it, or not being able to enter bootloader mode), i am working on it right now, so expect a fix asap 😄

**Comment by vicnil** [comment]
> > @AlexDygma I tried moving the Bazecor window but couldn't find anything behind it (or is it supposed to come up not at opening the Bazecor but at some other stage?). I'm on i3 windows manager, so it could be that it makes the pop-up harder for me to find. It will be interesting to test the new version!
> 
> Here you have it https://drive.google.com/file/d/1zij4I7AVAYjDa4wf1FnsNofPgQiL6B5h/view?usp=sharing

Yes, that is the version I'm using. I ran it too just in case the code is updated without the name being changed, but the window didn't pop up when starting Bazecor.

**Comment by nsrderooy** [comment]
@AlexDygma 
So it did not fix my problem, but I have found what the problem is.

Something is causing the udev rule to be ignored.
The permissions are not being set as defined in the rule.

The default permissions are;
```
crw-rw-rw- 1 root dialout 166, 0 Dec  6 12:53 /dev/ttyACM0
```

I added the dailout to my user and then it worked (I noticed that there was no permission in the inspect mode).

In case people want a quick fix;
```
sudo usermod -a -G dialout "${USERNAME}"
```

I'll have a look at the udev rule when I have time.

This is what happens when plugging in the keyboard;
![image](https://user-images.githubusercontent.com/6792480/144790126-95586cea-6e4b-46b0-abaa-21592ed247d0.png)


```
while true; do ls -l /dev/ttyACM0; done
```

**Comment by nsrderooy** [comment]
Found the issue.

The udev rule is being overwritten.

this is the fix;
```
sudo mv /etc/udev/rules.d/{50,10}-dygma.rules
sudo udevadm control --reload-rules
sudo udevadm trigger
```

check your usb device id by giving this command;
```
udevadm info $(grep -l Dygma /sys/bus/usb/devices/*/manufacturer | sed s/.manufacturer//g)
```
take the kast part of the `P: /devices/` line (in my case 7-1) and add it to this command
```
sudo udevadm test /sys/bus/usb/devices/7-1
```
Here you will notice that the Dygma rule is not the first line after `Loaded timestamp for '/etc/udev/rules.d'.`
lines above will overwrite your line.

I would say this is the solution, don't know if there is another way to overwrite the ordering.

**Comment by AlexDygma** [comment]
> Found the issue.
> 
> The udev rule is being overwritten.
> 
> **This is the fix;**
> 
> ```
> sudo mv /etc/udev/rules.d/{50,10}-dygma.rules
> sudo udevadm control --reload-rules
> sudo udevadm trigger
> ```
> 
> How I found the problem I have put below. Check your usb device id by giving this command;
> 
> ```
> udevadm info $(grep -l Dygma /sys/bus/usb/devices/*/manufacturer | sed s/.manufacturer//g)
> ```
> 
> Take the last part of the `P: /devices/` line (in my case 7-1) and add it to this command
> 
> ```
> sudo udevadm test /sys/bus/usb/devices/7-1
> ```
> 
> Here you will notice that the Dygma rule is not the first line after `Loaded timestamp for '/etc/udev/rules.d'.` lines above will overwrite your line.
> 
> I would say this is the solution, don't know if there is another way to overwrite the ordering.

Wow! great job pinning the problem! the thing is I don't really know why the uDev rules are being superseded in your linux setup, and I don't think it's so easy to automate, so do you know a way to give these rules more priority over the system ones? ill be looking into that myself also, if I find something I will post it here

**Comment by nsrderooy** [comment]
@AlexDygma sorry I was out for a bit (no access to my yubikey so I could not respond. :)
so you can supersede the rules by changing the leading number, your rule is set to 50, in my example I set it to 10, which is quite high, the `mv` command changes it to 10.
In the output of this command;
```
sudo udevadm test /sys/bus/usb/devices/7-1
```
it will show you the order it is loading the files in.
![image](https://user-images.githubusercontent.com/6792480/146635826-93826527-ba00-4878-b4db-6f0d5f8da82a.png)

here you can see that it is now being loaded before all other rules.

**Comment by nsrderooy** [comment]
@AlexDygma I see that this ticket is still open, did some more looking into it and the solution is to simply change the number;

> Rules Files
The udev rules are read from the files located in the system rules directories /usr/lib/udev/rules.d and /usr/local/lib/udev/rules.d, the volatile runtime directory /run/udev/rules.d and the local administration directory /etc/udev/rules.d. All rules files are collectively sorted and processed in lexical order, regardless of the directories in which they live. However, files with identical filenames replace each other. Files in /etc/ have the highest priority, files in /run/ take precedence over files with the same name under /usr/. This can be used to override a system-supplied rules file with a local file if needed; a symlink in /etc/ with the same name as a rules file in /usr/lib/, pointing to /dev/null, disables the rules file entirely. Rule files must have the extension .rules; other extensions are ignored.

source: https://www.freedesktop.org/software/systemd/man/udev.html

**Comment by nsrderooy** [comment]
@AlexDygma , I see that my proposal got merged (at least as per Bazecor-1.0.0) I would guess this ticket can be closed as such.
